routes/problems_id_solutions.js

```diff
@@ -0,0 +1,63 @@
+'use strict';
+
+const datastore             = require('../modules/datastore.js');
+const auth                  = require('../modules/auth.js');
+const accessControl         = require('../modules/access_control.js');
+
+const SOLUTION_KIND         = 'Solution';
+const DEFAULT_LIMIT         = 15;
+const DEFAULT_SORT          = 'complexity';
+const DEFAULT_IS_DESCENDING = false;
+
+
+/**
+ * Utility function to get a list of submitted solutions to display.
+ * Note that the code is not included in this query as it will be
+ * handled by POST request handler, which will check if the client
+ * has enough privilege to view the code.
+ **/
+async function getSolutions(problemId, limit, sortBy, isDescending) {
+  const query = datastore
+    .createQuery(SOLUTION_KIND)
+    .filter('problemId', '=', parseInt(problemId))
+    .order(sortBy, {
+      descending: isDescending,
+    })
+    .limit(parseInt(limit))
+    .select(['email', 'complexity']);
+
+  return (await datastore.runQuery(query))[0];
+}
+
+async function getSubmittedCode(problemId, email) {
+  const query = datastore
+    .createQuery(SOLUTION_KIND)
+    .filter('problemId', '=', parseInt(problemId))
+    .filter('email', '=', email)
+    .select(['code']);
+
+  return (await datastore.runQuery(query))[0];
+}
+
+module.exports = function(app) {
+  /**
+   * GET handler that displays problem dynamically according
+   * to the problem id given in the request parameter.
+   **/
+  app.get('/problems/:id/solutions', async function(request, response) {
+    const solutions = await getSolutions(request.params.id, DEFAULT_LIMIT, DEFAULT_SORT, DEFAULT_IS_DESCENDING);
+    response.render('solutions', {solutions: solutions});
+  });
+
+  app.post('/problems/:id/solutions', async function(request, response) {
```

Comment:
should be a get
https://google.aip.dev/131

Reply:
Thanks for the feedback!

How should we go about implementing access control if we use a GET request for this part?
We have been sending authentication token via request body but I've found online that:
```
As per RFC 7231

GET is the primary mechanism of information retrieval and the focus of almost all performance optimizations.

A payload within a GET request message has no defined semantics; sending a payload body on a GET request might cause some existing implementations to reject the request.
```
This led me to believe that the only way to transmit authentication token with GET request would be via request parameter, which would be unsafe.